File: frontend/ui_tests/pages/CopilotPage.ts
import { Page, expect } from "@playwright/test";

/**
 * Page Object Model for Copilot UI
 * Encapsulates all selectors and interactions for the copilot page.
 */
export class CopilotPage {
  readonly page: Page;
  readonly baseUrl: string;

  // Selectors
  readonly modeToggleData = 'button:has-text("Data")';
  readonly modeToggleAIReports = 'button:has-text("AI Reports")';
  readonly modeToggleInsights = 'button:has-text("Insights")';
  readonly queryInput =
    'input[placeholder*="search"], input[placeholder*="query"], textarea, input[type="text"]';
  readonly askButton =
    'button:has-text("Ask"), button:has-text("Send"), button:has-text("Submit")';
  readonly resultsContainer =
    '[data-testid="results"], .results, .output, #results, .response';
  readonly resultTable = "table";
  readonly resultCard = '.card, [data-testid="result-card"], .card-body';
  readonly aiBadge = ':has-text("AI"), :has-text("Generated"), :has-text("Report")';
  readonly errorMessage = '.error, [role="alert"], .alert';

  constructor(page: Page) {
    this.page = page;
    this.baseUrl = "http://localhost:8000";
  }

  /**
   * Navigate to the copilot page
   */
  async navigate() {
    await this.page.goto(`${this.baseUrl}/`);
  }

  /**
   * Wait for page to be fully loaded
   */
  async waitForPageLoad(timeout = 5000) {
    await this.page.waitForLoadState("networkidle", { timeout });
  }

  /**
   * Select "Data" mode
   */
  async selectDataMode() {
    const buttons = await this.page.locator("button").all();
    for (const btn of buttons) {
      const text = await btn.textContent();
      if (text && text.toLowerCase().includes("data")) {
        await btn.click();
        return;
      }
    }
  }

  /**
   * Select "AI Reports" mode
   */
  async selectAIReportsMode() {
    const buttons = await this.page.locator("button").all();
    for (const btn of buttons) {
      const text = await btn.textContent();
      if (
        text &&
        (text.toLowerCase().includes("ai") ||
          text.toLowerCase().includes("report"))
      ) {
        await btn.click();
        return;
      }
    }
  }

  /**
   * Select "Insights" mode
   */
  async selectInsightsMode() {
    const buttons = await this.page.locator("button").all();
    for (const btn of buttons) {
      const text = await btn.textContent();
      if (text && text.toLowerCase().includes("insight")) {
        await btn.click();
        return;
      }
    }
  }

  /**
   * Enter query text in the input field
   */
  async enterQuery(queryText: string) {
    // Find text input or textarea
    const inputs = await this.page
      .locator(
        'input[type="text"], textarea, input:not([type]), input[placeholder]'
      )
      .all();

    if (inputs.length > 0) {
      await inputs[0].fill(queryText);
    } else {
      // Fallback: find any input
      const anyInput = this.page.locator("input").first();
      await anyInput.fill(queryText);
    }
  }

  /**
   * Click the "Ask" or "Send" button
   */
  async clickAskButton() {
    const buttons = await this.page.locator("button").all();
    for (const btn of buttons) {
      const text = await btn.textContent();
      if (
        text &&
        (text.toLowerCase().includes("ask") ||
          text.toLowerCase().includes("send") ||
          text.toLowerCase().includes("submit"))
      ) {
        await btn.click();
        return;
      }
    }
    // Fallback: click first button if none matched
    if (buttons.length > 0) {
      await buttons[0].click();
    }
  }

  /**
   * Get the visible results text
   */
  async getResultsText(): Promise<string> {
    try {
      const text = await this.page.locator("body").textContent();
      return text || "";
    } catch {
      return "";
    }
  }

  /**
   * Check if results are displayed
   */
  async isResultsVisible(): Promise<boolean> {
    try {
      const selector = this.page.locator(
        '[data-testid="results"], .results, .output, #results, .response, table, .card'
      );
      const count = await selector.count();
      return count > 0;
    } catch {
      return false;
    }
  }

  /**
   * Check if results table is visible
   */
  async isTableVisible(): Promise<boolean> {
    try {
      const table = this.page.locator(this.resultTable);
      return (await table.count()) > 0;
    } catch {
      return false;
    }
  }

  /**
   * Check if AI badge or AI content is visible
   */
  async isAIBadgeVisible(): Promise<boolean> {
    try {
      const badges = await this.page
        .locator(
          'text="AI", text="Generated", text="Report", text="Analysis"'
        )
        .all();
      return badges.length > 0;
    } catch {
      return false;
    }
  }

  /**
   * Get error message if displayed
   */
  async getErrorMessage(): Promise<string | null> {
    try {
      const error = await this.page.locator(this.errorMessage).textContent();
      return error;
    } catch {
      return null;
    }
  }

  /**
   * Wait for results to appear on page
   */
  async waitForResults(timeout = 10000) {
    // Wait for either results container, table, or any visible content
    try {
      await this.page
        .locator(
          "[data-testid=\"results\"], .results, .output, #results, table, .card"
        )
        .first()
        .waitFor({ state: "visible", timeout });
    } catch {
      // Even if specific selector not found, page may have rendered content
    }
  }

  /**
   * Set up route interception for API calls (call before navigate)
   */
  async setupRouteInterception() {
    // Intercept /copilot/ask
    await this.page.route("**/copilot/ask", async (route) => {
      await route.fulfill({
        status: 200,
        contentType: "application/json",
        body: JSON.stringify({
          intent: "list_purchase_orders",
          message: "Found 10 purchase orders.",
          data: [
            {
              name: "PO-2024-001",
              supplier: "Supplier A",
              grand_total: 5000.0,
              status: "Completed",
            },
            {
              name: "PO-2024-002",
              supplier: "Supplier B",
              grand_total: 7500.0,
              status: "Submitted",
            },
          ],
        }),
      });
    });

    // Intercept /ai/report
    await this.page.route("**/ai/report", async (route) => {
      await route.fulfill({
        status: 200,
        contentType: "application/json",
        body: JSON.stringify({
          report:
            "Monthly procurement analysis shows stable supplier performance.",
          summary: "Cost-effective purchasing with competitive rates.",
        }),
      });
    });
  }
}

File: frontend/ui_tests/pages/copilot_page.py
"""
Page Object Model for Copilot UI.
Encapsulates all selectors and interactions for the copilot page.
"""

from playwright.sync_api import Page
from .base_page import BasePage


class CopilotPage(BasePage):
    """Page Object for the main Copilot UI page."""

    # Selectors as class constants
    MODE_BUTTON_DATA = "button"  # Will use get_by_role
    ASK_BUTTON = "button"  # Will use get_by_role
    QUERY_INPUT_PLACEHOLDER = "ask"
    RESULTS_CONTAINER = ".results"
    RESULT_TABLE = "table"
    ERROR_MESSAGE = "[role='alert']"
    EXPORT_BUTTON = "button"  # Will use get_by_role

    def __init__(self, page: Page, base_url: str = "http://127.0.0.1:8000") -> None:
        """
        Initialize Copilot page object.

        Args:
            page: Playwright Page instance
            base_url: Base URL for the application
        """
        super().__init__(page, base_url)

    def navigate_to_copilot(self) -> None:
        """Navigate to the copilot page."""
        self.navigate("/")
        # Wait for page to load (use load instead of networkidle to avoid hanging)
        self.page.wait_for_load_state("load", timeout=5000)

    def select_data_mode(self) -> None:
        """Click the 'Data' mode button."""
        try:
            self.click_by_role("button", "Data")
        except Exception:
            # Button might not exist or be already selected
            pass

    def select_ai_reports_mode(self) -> None:
        """Click the 'AI Reports' mode button."""
        try:
            self.click_by_role("button", "AI Reports")
        except Exception:
            pass

    def select_insights_mode(self) -> None:
        """Click the 'Insights' mode button."""
        try:
            self.click_by_role("button", "Insights")
        except Exception:
            pass

    def ask_question(self, query: str, wait_for_response: bool = True) -> None:
        """
        Write a question and click ask.

        Args:
            query: Question to ask
            wait_for_response: Wait for response to appear
        """
        # Fill the query input
        try:
            # Try to find input by placeholder
            self.page.get_by_placeholder("ask", exact=False).fill(query)
        except Exception:
            # Fallback: find any text input and fill it
            inputs = self.page.locator("input[type='text'], textarea")
            if inputs.count() > 0:
                inputs.first.fill(query)

        # Click ask button
        try:
            self.click_by_role("button", "Ask")
        except Exception:
            self.click_by_text("Ask")

        # Wait for response if requested
        if wait_for_response:
            self.wait_for_response()

    def wait_for_response(self, timeout: int = 10000) -> None:
        """
        Wait for response to appear after asking.

        Args:
            timeout: Timeout in milliseconds
        """
        self.page.wait_for_timeout(2000)  # Give API time to respond

    def get_results_text(self) -> str:
        """
        Get the text content of results.

        Returns:
            Results text content
        """
        content = self.get_page_content()
        return content if content else ""

    def has_results(self) -> bool:
        """
        Check if results are displayed.

        Returns:
            True if results visible
        """
        content = self.get_results_text()
        return any(keyword in content.lower() for keyword in ["supplier", "item", "data", "result"])

    def has_error(self) -> bool:
        """
        Check if error message is displayed.

        Returns:
            True if error visible
        """
        return self.is_visible(self.ERROR_MESSAGE)

    def get_error_message(self) -> str | None:
        """
        Get error message text.

        Returns:
            Error message or None
        """
        if self.has_error():
            return self.get_text_content(self.ERROR_MESSAGE)
        return None

    def find_export_button(self) -> bool:
        """
        Look for export/download button.

        Returns:
            True if export button found
        """
        buttons = self.page.locator("button")
        for i in range(buttons.count()):
            button = buttons.nth(i)
            text = button.text_content() or ""
            if any(keyword in text.lower() for keyword in ["export", "download", "csv", "save"]):
                return True
        return False

    def click_export_button(self) -> bool:
        """
        Find and click export button.

        Returns:
            True if button was clicked
        """
        buttons = self.page.locator("button")
        for i in range(buttons.count()):
            button = buttons.nth(i)
            text = button.text_content() or ""
            if any(keyword in text.lower() for keyword in ["export", "download", "csv"]):
                button.click()
                return True
        return False

    def get_table_data(self) -> str | None:
        """
        Get table content if present.

        Returns:
            Table text content or None
        """
        if self.page.locator(self.RESULT_TABLE).count() > 0:
            return self.get_text_content(self.RESULT_TABLE)
        return None

    def is_page_loaded(self) -> bool:
        """
        Check if page is fully loaded.

        Returns:
            True if page is loaded
        """
        try:
            self.wait_for_load_state("networkidle", timeout=3000)
            return True
        except Exception:
            return False
